src/dune_rules/dune_env.ml

```diff
@@ -226,10 +226,27 @@ module Stanza = struct
        and+ configs = fields config in
        (pat, configs))
 
+  let check_rules ~version ~loc rules =
+    let rec has_after_any = function
+      | [ (Any, _) ] -> false
```

Comment:
Maybe you could merge the `[]` base case with that one ?
```suggestion
      | [ (Any, _) ] | [] -> false
```

(You could also remove this first case and just use a catch-all pattern but I feel like it is more readable with the expanded cases.)